### PR TITLE
Fix Marinade settlement funding detection in verify_settlement

### PR DESCRIPTION
**Pull request by martinkrecek**
## Summary

Fix a critical bug in `verify_settlement` where Marinade-funded settlements were incorrectly flagged as non-funded. The issue occurs because Marinade settlements are funded by creating stake accounts assigned to the settlement's staker authority directly, without calling the `fund_settlement` instruction. This means `Settlement.lamports_funded` remains 0 forever for Marinade settlements, even when fully funded and claimed.

## Key Changes

- **verify_settlement.rs**: Updated settlement funding detection logic to recognize three valid funding states:
  1. `lamports_funded > 0` (ValidatorBond path via `fund_settlement` instruction)
  2. `lamports_claimed > 0` (settlement already claimed, implying it was funded)
  3. Stake accounts assigned to the settlement's staker authority hold sufficient lamports (Marinade path)

- **stake_accounts.rs**: Added `settlement_funded_claimable_lamports()` function to calculate claimable lamports from stake accounts assigned to a settlement's staker authority, accounting for the minimum buffer (`minimal_stake_lamports`) that must remain in each account.

- **fund_settlement.rs**: Updated funding amount calculation to use `settlement_funded_claimable_lamports()` for Marinade settlements, making the funding process idempotent (prevents duplicate stake account creation on repeated runs).

- **Tests**: Added comprehensive test coverage for:
  - Marinade-funded settlements (claimed and unclaimed)
  - ValidatorBond-funded settlements
  - Underfunded settlements (still correctly flagged)
  - Edge cases (dust accounts, no matching stake accounts, saturation)

## Implementation Details

The fix mirrors how the on-chain `fund_settlement` and `close_settlement` instructions load stake accounts — by querying for accounts whose staker authority matches the settlement's staker authority. For Marinade settlements, this is the only on-chain source of truth for funding status since `lamports_funded` is never bumped.

The `minimal_stake_lamports` buffer (minimum delegation + rent exemption) is subtracted per account to account for the minimum amount that must remain to keep the stake account valid after claims drain it.

https://claude.ai/code/session_01JM7GC7A3LY4GPY8W5Tb9Tu

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved settlement funding calculations for Marinade- and ValidatorBond-funded settlements.
  * Prevented repeated funding runs from incorrectly creating duplicate stake accounts.
  * Improved settlement verification to recognize funding reflected through claimed or stake-backed balances.
  * Reduced false reports of unfunded Marinade settlements while continuing to identify genuinely unfunded or underfunded settlements.
  * Added safeguards for small residual balances to ensure accurate funding status without calculation errors.

### CODE REVIEW

File: settlement-pipelines/src/stake_accounts.rs [hunk 1]
```diff
@@ -232,6 +232,33 @@ pub fn filter_settlement_funded(
...
+/// The on-chain `Settlement.lamports_funded` field is increased only by the `fund_settlement`
+/// instruction (the ValidatorBond funding path). Marinade-funded settlements are funded by
+/// creating a stake account assigned to the settlement staker authority directly, without calling
+/// `fund_settlement`, so their `lamports_funded` stays `0` forever. For those settlements the
+/// lamports held by the stake accounts under the settlement staker authority are the only on-chain
+/// source of truth for how much has been funded.
+///
+/// Each funding stake account must keep `minimal_stake_lamports` (minimum delegation + rent
+/// exemption) to remain a valid stake account once claims have drained it, so that buffer is
+/// excluded per account — the remainder is what is actually claimable.
```
[the remark this answers is not in the document]
Comment (martinkrecek) [reply settlement-pipelines/src/stake_accounts.rs]: ok, updating the comment

**Comment by kronael** [comment]
We can also just remove Marinade Settlements altogether... I'll be working on this some time later this week...

File: settlement-pipelines/src/bin/fund_settlement.rs [hunk 2]
```diff
@@ -271,22 +271,40 @@ async fn prepare_funding(
...
-                assert_eq!(
-                    settlement.max_total_claim,
-                    settlement_record.max_total_claim_sum,
-                );
-                settlement
-                    .max_total_claim
-                    .saturating_sub(settlement.lamports_funded)
-            },
-        );
+        // `Settlement.lamports_funded` is increased only by the on-chain `fund_settlement`
```
Comment (ochaloup) [inline settlement-pipelines/src/bin/fund_settlement.rs]: Maybe to shorten this lengthy comment a bit?

>   // Marinade funds a settlement by creating a stake account directly (no `fund_settlement`
>  // call), so `lamports_funded` stays 0 — deriving amount_to_fund from it would re-fund and
>  // duplicate the stake account every run. Reconstruct: claimed + claimable held in stake accounts.

File: settlement-pipelines/src/bin/verify_settlement.rs [hunk 3]
```diff
@@ -155,7 +162,25 @@ fn verify_epoch_settlements(
             if let Some(onchain_settlement) =
                 onchain_settlements.get(&listed_settlement.settlement_address)
             {
-                if onchain_settlement.lamports_funded == 0 {
+                // A settlement is funded if the on-chain `fund_settlement` instruction has run
```
Comment (ochaloup) [inline settlement-pipelines/src/bin/verify_settlement.rs]: Maybe shorten the comment a bit?
e.g., 

>  // Marinade funding never bumps `lamports_funded`; detect it via stake accounts assigned to the
> // settlement staker authority. Funded if lamports_funded>0, or claimed+claimable covers
 > // max_total_claim (funds are conserved, so this holds at any claim stage).

File: settlement-pipelines/src/bin/fund_settlement.rs [hunk 4]
```diff
@@ -271,22 +271,40 @@ async fn prepare_funding(
...
+        // instruction, which is the ValidatorBond funding path. Marinade-funded settlements are
+        // funded by creating a stake account assigned to the settlement staker authority directly
+        // (no `fund_settlement` call), so their `lamports_funded` stays 0 forever. Deriving
+        // `amount_to_fund` from `lamports_funded` would therefore make Marinade funding
+        // non-idempotent: every run recomputes the full amount and creates yet another duplicate
+        // stake account. For the Marinade funder we reconstruct how much is already funded from
+        // on-chain reality instead — lamports already claimed plus claimable lamports still held by
+        // the settlement's stake accounts.
+        let already_funded = match &settlement_record.funder {
+            SettlementFunderType::Marinade(_) => {
```
Comment (ochaloup) [inline settlement-pipelines/src/bin/fund_settlement.rs:279]: We have been asserting on-chain `max_total_claim == JSON max_total_claim_sum` before in the code, could that be added here into arm as well or it's not valid anymore from your perspective?

**Review by ochaloup — DISMISSED** [review]
Thanks 🤟

File: settlement-pipelines/src/bin/fund_settlement.rs (hunk 4 above)
Comment (martinkrecek) [reply settlement-pipelines/src/bin/fund_settlement.rs:279]: good point 👍 added

**Review by ochaloup — APPROVED** [review]
Thank you!
